pipeline_steps/common.groovy

```diff
@@ -1969,3 +1974,144 @@ void updateStringParam(String name, String value, String description){
+  return startTime.plus(getPredictedDuration())
+}
+
+/**
+* Get date object representing the start of the next maintenance window
+*/
+LocalDateTime getNextMaintenanceWindowStart(){
+  // set start of maintenance window
+
+  DayOfWeek maintDay = DayOfWeek.MONDAY
+  Integer maintHour = 10
+  Integer daysInWeek = 7
+
+  // get current time & date
+  LocalDateTime now = LocalDateTime.now()
+  LocalDateTime maintStart
+
+  if (now.getDayOfWeek() == maintDay){
+    // its Monday, we either need to return today's maintenance window
+    // or next Monday's.
+    if (now.getHour() > maintHour){
+      // return next week's maintenance window
+      maintStart = now.plusDays(7)
+    } else {
+      maintStart = now
+    }
+  } else {
+    // Its not Monday.
+    Integer doW = now.getDayOfWeek().getValue()
+    Integer mDoW = maintDay.getValue()
+    Integer daysTillNextMaint = (daysInWeek -(doW - mDoW)) % daysInWeek
+    maintStart = now.plusDays(daysTillNextMaint)
+  }
+
+  // Maintstart is now a date time with the correct date, fix the time and return it.
+  return maintStart.withHour(maintHour)\
+                   .withMinute(0)\
+                   .withSecond(0)\
+                   .withNano(0)
+
+}
+
+/**
+* Check if this job is likely to run into the next maintenance window
+*/
+Boolean willOverlapMaintenanceWindow(){
+  // The pr buffer is an hour before the maintenance window starts.
+  // PRs must be predicted to finish before the PR buffer starts
+  // to be allowed to run. This gives PR jobs an hour to overrun
+  // their predicted duration.
+  Duration prBuffer = Duration.ofHours(1)
+
+  LocalDateTime completion = getPredictedCompletionTime()
+  LocalDateTime maintWindowStart = getNextMaintenanceWindowStart()
+  LocalDateTime prBufferStart = maintWindowStart.minus(prBuffer)
+
+  return prBufferStart.isBefore(completion)
+}
+
+Boolean issueExistsForNextMaintenanceWindow(){
+  LocalDateTime nextWindow = getNextMaintenanceWindowStart()
+  String nextWindowDateString = nextWindow.format(DateTimeFormatter.ISO_LOCAL_DATE)
+  List issues = jira_query("project=RE AND summary ~ '${nextWindowDateString} Maintenance Window'")
+  return issues.size > 0
+}
+
+/*
+* Create an issue for the next maintenance window, returns the issue key
+*/
+String getOrCreateIssueForNextMaintenanceWindow(){
+  LocalDateTime nextWindow = getNextMaintenanceWindowStart()
+  String nextWindowDateString = nextWindow.format(DateTimeFormatter.ISO_LOCAL_DATE)
+  String jiraIssue = get_or_create_jira_issue(
+    "RE",
+    "BACKLOG",
+    "${nextWindowDateString} Maintenance Window",
+    "Jira issue to collect notes relating to a maintenance window, for more information please see https://rpc-openstack.atlassian.net/wiki/spaces/RE/pages/469794817/RE+Infrastructure+Maintenance+Window",
+    ["jenkins", "re", "maintenance"]
+  )
+  return jiraIssue
+}
+
+void abortIfMaintenanceWindowPending(){
+  if (issueExistsForNextMaintenanceWindow() && willOverlapMaintenanceWindow()){
```

Comment:
It would be nice if this allowed an ad-hoc maintenance to be scheduled outside of the normal window.

Reply:
Created RE-1958 for this